### PR TITLE
Add ability to select multiple rows on Admin Tables

### PR DESCRIPTION
**Pull request by DanielePalombo**
**Description**

This PR adds the ability to select more elements of admin tables.
Before this PR the ability was present only for return items and with a [custom js code](https://github.com/solidusio/solidus/blob/6a21551c8900eb899b2e26fa2efd8876c2293eed/backend/app/assets/javascripts/spree/backend/returns/return_item_selection.js#L24-L28).

Adding a `selectable-table` CSS class to an admin table, and a checkbox (with a `selectable` class) for every related record, it will be possible to select more than one element.

Screenshot 2020-03-20 at 16 47 39

If a checkbox named `select-all` is present this check/uncheck all the elements.

Screenshot 2020-03-20 at 16 52 37

Adding the selector CSS class `selectable-label` on a node HTML it will
be rendered a text depending on the number of the selected row.

Thanks to it will be possible to show different information depending on the selection type.
e.g.
Screenshot 2020-03-20 at 15 49 27
Screenshot 2020-03-20 at 15 48 59

An example of usage is provided at `/admin/style_guide` under `Building Tables` section.

This PR uses the selectable code on the return items

**Checklist:**
- [X] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [X] I have added a detailed description into each commit message
- [X] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [X] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: backend/app/assets/javascripts/spree/backend/components/selectable_table.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+//= require solidus_admin/Sortable
+/* eslint no-unused-vars: "off" */
```
Comment (kennyadsl) [inline backend/app/assets/javascripts/spree/backend/components/selectable_table.js:2]: Do we need this?

File: backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb [hunk 2]
```diff
@@ -1,7 +1,19 @@
...
+  <table class='selectable-table'>
     <thead>
+      <tr class="selectable-header no-item-selected">
+        <th><%= check_box_tag 'select-all' %></th>
+        <th colspan="7">
+          <span class="all-selected">
+            All Item selected
+          </span>
+          <span class="custom-selection">
+            <span class="label"></span> Selected Item
```
Comment (kennyadsl) [inline backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb:11]: Why `label`? Do you think we can use a better name for this class to describe its content?

File: backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb [hunk 3]
```diff
@@ -12,6 +24,7 @@
 
     <tbody>
       <tr>
+        <td><%= check_box_tag 'selected_ids[]', 1, false, class: 'selectable' %>
```
Comment (kennyadsl) [inline backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb:27]: Does this miss the closing `</td>` tag?

File: backend/app/assets/javascripts/spree/backend/components/selectable_table.js (hunk 1 above)
Comment (DanielePalombo) [reply backend/app/assets/javascripts/spree/backend/components/selectable_table.js:2]: Nope.

File: backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb (hunk 2 above)
Comment (DanielePalombo) [reply backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb:11]: `selected-item-label` ?

File: backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb (hunk 3 above)
Comment (DanielePalombo) [reply backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb:27]: 👍

File: backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb (hunk 2 above)
Comment (kennyadsl) [reply backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb:11]: What will be the content of that span? It's a number right? `selected-item-count`?

Comment (DanielePalombo) [reply backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb]: You are right, it's a number! I'm going to change it

File: backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb [hunk 4]
```diff
@@ -1,7 +1,19 @@
 <%- snippet = capture do %>
-  <table>
+  <table class='selectable-table'>
     <thead>
+      <tr class="selectable-header no-item-selected">
+        <th><%= check_box_tag 'select-all' %></th>
+        <th colspan="7">
+          <span class="all-selected">
```
Comment (kennyadsl) [inline backend/app/views/spree/admin/style_guide/topics/tables/_building_tables.html.erb:7]: As I said IRL, I'd move this part in an hbs template.

**Comment by kennyadsl** [comment]
We discussed IRL, and we are trying to move more things to handlebars templates.

**Comment by DanielePalombo** [comment]
Refactored with handlebars templates
Exported return items selection logic in a Backbone controller and handlebars template

File: backend/app/assets/javascripts/spree/backend/views/tables/selectable_table_sum_item_amount.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+Spree.Views.Tables.SelectableTable.SumItemAmount = Backbone.View.extend({
+  initialize: function(options) {
+    this.listenTo(this.model, 'change', this.render)
+
+    this.render();
+  },
+
+  render: function() {
+    var html = HandlebarsTemplates['tables/return_item_sum_amount']({
```
Comment (kennyadsl) [inline backend/app/assets/javascripts/spree/backend/views/tables/selectable_table_sum_item_amount.js:9]: This file is both generic (it's called with a generic `SumItemAmount` name) and specific since it only works with this return items table. I'd say we can improve it by passing dynamic data (like the template and the label to use) from outside. Do you think it's doable?

Comment (DanielePalombo) [reply backend/app/assets/javascripts/spree/backend/views/tables/selectable_table_sum_item_amount.js]: I refactored the code. I think we can add the ability to customize the method to recalculate the item amount.
Right now the commit creates a specific object to wrap the generic object and it provides to the generic object these data.

**Comment by kennyadsl** [comment]
@DanielePalombo Not sure how, but the failure doesn't seem a flaky, can you take a look? I'm restarting the suite again just be sure anyway.

**Comment by DanielePalombo** [comment]
@kennyadsl It looks green now

File: backend/app/assets/javascripts/spree/backend/components/selectable_table.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+Spree.ready(function() {
+  $('table.selectable-table').each(function() {
+    var selectableTableModel = new Backbone.Model({
+      allSelected: false,
+      selectedItems: []
+    })
+
+    new Spree.Views.Tables.SelectableTable({el: this, model: selectableTableModel});
+
+    if($(this).hasClass('return-items-table')) {
```
Comment (tvdeyen) [inline backend/app/assets/javascripts/spree/backend/components/selectable_table.js:10]: Why hard-code this behavior here? Can we add a dedicated view model for the return items table instead?

File: backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<th>
+  <input type="checkbox" {{#if all_items_selected}}checked="checked"{{/if}} name="select-all" class="selectAll" id="select-all" value="1">
+</th>
+<th colspan="{{colspan}}">
+  {{#if all_items_selected}}
+    All Items Selected
```
Comment (tvdeyen) [inline backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs:6]: This needs to be translatable

File: backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  {{else}}
+    {{#if no_item_selected}}
+      No Items Selected
```
Comment (tvdeyen) [inline backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs:9]: ditto

File: backend/app/assets/stylesheets/spree/backend/sections/_style_guide.scss [hunk 9]
```diff
@@ -122,6 +122,10 @@
 .style-guide-code,
 .style-guide-result {
   margin: 1rem 0;
+
+  img {
+    max-width: none;
+  }
```
Comment (tvdeyen) [inline backend/app/assets/stylesheets/spree/backend/sections/_style_guide.scss:128]: Why is this change necessary?

File: backend/app/assets/javascripts/spree/backend/components/selectable_table.js (hunk 6 above)
Comment (spaghetticode) [inline backend/app/assets/javascripts/spree/backend/components/selectable_table.js:10]: I wonder if this code could be moved in a location more specific to return items. This file seems to address the generic behavior for selectable tables, so it's a bit strange seeing something very specific related only to the return items table... what do you think?

Comment (spaghetticode) [reply backend/app/assets/javascripts/spree/backend/components/selectable_table.js:10]: Sorry, I just realized that @tvdeyen made the same comment some minutes ago. I'll let you handle that discussion with him

File: backend/app/assets/stylesheets/spree/backend/sections/_style_guide.scss (hunk 9 above)
Comment (DanielePalombo) [reply backend/app/assets/stylesheets/spree/backend/sections/_style_guide.scss:128]: This shows the image under style_guide page
Screenshot 2020-06-26 at 16 35 22

Without it, the image doesn't appear
Screenshot 2020-06-26 at 16 35 06

File: backend/app/assets/javascripts/spree/backend/components/selectable_table.js (hunk 6 above)
Comment (DanielePalombo) [reply backend/app/assets/javascripts/spree/backend/components/selectable_table.js]: I made a dedicated view for return items table, let me know if it works for you.

File: backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs (hunk 7 above)
Comment (DanielePalombo) [reply backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs]: Done

File: backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs (hunk 8 above)
Comment (DanielePalombo) [reply backend/app/assets/javascripts/spree/backend/templates/tables/selectable_label.hbs]: done

File: backend/app/assets/javascripts/spree/backend/views/tables/return_items.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+Spree.Views.Tables.ReturnItems = Backbone.View.extend({
+  initialize: function() {
+    if(this.$el.hasClass('return-items-table')) {
+      var tfoot = document.createElement('tfoot')
+      new Spree.Views.Tables.SelectableTable.SumReturnItemAmount({ el: tfoot, model: this.model});
```
Comment (kennyadsl) [inline backend/app/assets/javascripts/spree/backend/views/tables/return_items.js:5]: Please use the same spacing rule for starting and ending inside `{}`

Comment (DanielePalombo) [reply backend/app/assets/javascripts/spree/backend/views/tables/return_items.js]: Done!

**Comment by DanielePalombo** [comment]
@kennyadsl the build is failed because of flaky spec, do you mind re-running the CI?
